openedx/features/enterprise_support/utils.py

```diff
@@ -293,42 +294,75 @@ def _get_sync_learner_profile_data(enterprise_customer):
     return False
 
 
-def get_enterprise_learner_portals(request):
+def get_enterprise_learner_portal(request):
     """
-    Gets the formatted portal names and slugs that can be used
-    to generate links for enabled enterprise Learner Portals.
+    Gets the formatted portal name and slug that can be used
+    to generate a link for an enabled enterprise Learner Portal.
 
-    Caches and returns results in/from the user's request session if provided.
+    Caches and returns result in/from the user's request session if provided.
     """
     # Prevent a circular import.
-    from openedx.features.enterprise_support.api import enterprise_enabled
+    from openedx.features.enterprise_support.api import enterprise_enabled, enterprise_customer_uuid_for_request
 
     user = request.user
     # Only cache this if a learner is authenticated (AnonymousUser exists and should not be tracked)
+
+    learner_portal_session_key = 'enterprise_learner_portal'
+
     if enterprise_enabled() and ENTERPRISE_HEADER_LINKS.is_enabled() and user and user.id:
         # If the key exists return that value
-        if 'enterprise_learner_portals' in request.session:
-            return json.loads(request.session['enterprise_learner_portals'])
-        # Ordering is important, this is consistent with how we decide on which
-        # enterprise_customer is the selected one for an enterprise_customer
-        enterprise_learner_portals = [{
-            'name': enterprise_customer_user.enterprise_customer.name,
-            'slug': enterprise_customer_user.enterprise_customer.slug,
-            'logo': enterprise_customer_user.enterprise_customer.branding_configuration.logo.url if
-            enterprise_customer_user.enterprise_customer.branding_configuration else None,
-        } for enterprise_customer_user in EnterpriseCustomerUser.objects.filter(
-            user_id=user.id, enterprise_customer__enable_learner_portal=True
-        ).prefetch_related(
-            'enterprise_customer', 'enterprise_customer__branding_configuration'
-        ).order_by('-enterprise_customer__active', '-modified')]
+        if learner_portal_session_key in request.session:
+            return json.loads(request.session[learner_portal_session_key])
 
-        # Cache the result in the user's request session
-        request.session['enterprise_learner_portals'] = json.dumps(enterprise_learner_portals)
+        kwargs = {
+            'user_id': user.id,
+            'enterprise_customer__enable_learner_portal': True,
+        }
+        enterprise_customer_uuid = enterprise_customer_uuid_for_request(request)
+        if enterprise_customer_uuid:
+            kwargs['enterprise_customer__uuid'] = enterprise_customer_uuid
+
+        queryset = EnterpriseCustomerUser.objects.filter(**kwargs).prefetch_related(
+            'enterprise_customer',
+            'enterprise_customer__branding_configuration',
```

Comment:
Just to confirm, I'm pretty positive that `prefetch_related` shouldn't break on a null related field, but I saw these lines mentioned as a possible cause of the branding config errors so wanted to check

Reply:
Yeah, that's covered by the try/except in `enterprise_branding_configuration()` below.  I couldn't think of a way to have the Django ORM automatically give me a null if a related 1-1 model doesn't exist.  It's not the prefetch that fails, FWIW, it's referencing the related field on a model instance when the related field does not exist.  The prefetch is irrelevant to that error.